api/handlers.go

```diff
@@ -882,8 +882,381 @@ func (a *App) handleUnsubscribe(c *gin.Context) {
 	concludeRequest(c, nil, err)
 }
 
+
+//Bookmark handlers
+func (a *App) handleGetBookmarks(c *gin.Context) {
+	var r GetBookmarksRequest
+	if c.Bind(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+
+	mods := []qm.QueryMod{models.BookmarkWhere.UserID.EQ(user.ID)}
+
+	total, err := models.Bookmarks(mods...).Count(db)
+
+	mods = append(mods, qm.Load(models.BookmarkRels.BookmarkFolders))
+	_, offset := appendListMods(&mods, r.ListRequest)
+	if int64(offset) >= total {
+		concludeRequest(c, new(GetBookmarksResponse), nil)
+		return
+	}
+	appendQueryFilter(&mods, r.QueryFilter, "name")
+	if len(r.FolderIDsFilter) > 0 {
+		mods = append(mods,
+			qm.InnerJoin("bookmark_folder bf ON id = bf.bookmark_id"),
+			qm.WhereIn("bf.folder_id in ?", utils.ConvertArgsInt64(r.FolderIDsFilter)...),
+		)
+	}
+
+	bookmarks, err := models.Bookmarks(mods...).All(db)
+	if err != nil {
+		errs.NewInternalError(pkgerr.WithStack(err)).Abort(c)
+		return
+	}
+
+	items := make([]*Bookmark, len(bookmarks))
+	for i, b := range bookmarks {
+		items[i] = makeBookmarkDTO(b)
+	}
+
+	resp := GetBookmarksResponse{
+		ListResponse: ListResponse{Total: total},
+		Items:        items,
+	}
+	concludeRequest(c, resp, err)
+}
+
+func (a *App) handleCreateBookmark(c *gin.Context) {
+	var r AddBookmarksRequest
+	if c.BindJSON(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+	bookmark := &models.Bookmark{
+		UserID:     user.ID,
+		SourceUID:  r.SourceUID,
+		SourceType: r.SourceType,
+	}
+	if r.Name != "" {
+		bookmark.Name = null.StringFrom(r.Name)
+	}
+	if r.Data != nil {
+		data, err := json.Marshal(r.Data)
+		bookmark.Data = null.JSONFrom(data)
```

Comment:
no reason to have this before the error check one line below

Reply:
moved